### PR TITLE
GSYE-205: Enable residual requirements when DoF is activated

### PR DESCRIPTION
**Pull request by Hassan754**
**INTEGRATION_TESTS_BRANCH**=master

### CODE REVIEW

File: src/gsy_e/models/market/two_sided.py [hunk 1]
```diff
@@ -440,13 +441,28 @@ def _replace_offers_bids_with_residual_in_recommendations_list(
...
         """
 
+        def _adapt_matching_requirements_in_residuals(recommendation):
+            if "energy" in (recommendation.get("matching_requirements") or {}).get(
+                    "bid_requirement", {}):
+                for index, requirement in enumerate(recommendation["bid"]["requirements"]):
+                    if requirement == recommendation["matching_requirements"]["bid_requirement"]:
+                        bid_requirement = deepcopy(requirement)
+                        bid_requirement["energy"] -= bid_trade.traded_energy
+                        recommendation["bid"]["requirements"][index] = bid_requirement
```
Comment (hannesdiedrich) [inline src/gsy_e/models/market/two_sided.py:451]: Just out of interest: The garbage collector will delete the "old" requirement , right?

File: tests/test_two_sided_market.py [hunk 2]
```diff
@@ -439,6 +439,40 @@ def test_matching_list_gets_updated_with_residual_offers():
...
+                          traded_energy=1, trade_price=0.5)
+        recommendations = (
+            TwoSidedMarket._replace_offers_bids_with_residual_in_recommendations_list(
+                recommendations, offer_trade, bid_trade))
+        assert len(recommendations) == 1
+        assert recommendations[0]["offer"]["id"] == "residual_offer"
+        assert recommendations[0]["bid"]["id"] == "residual_bid"
+        # The bid traded 1 kWh so the maximum energy should become 2 - 1 = 1
+        # Furthermore, the bid was traded with the trading_partner
+        # it only bought 1 out of 1.5, so the residual bid's requirement should reflect this = 0.5
```
Comment (hannesdiedrich) [inline tests/test_two_sided_market.py:471]: ❤️

File: src/gsy_e/models/market/grid_fees/constant_grid_fees.py [hunk 3]
```diff
@@ -52,7 +52,7 @@ def update_forwarded_bid_trade_original_info(self, trade_original_info, market_b
             return None
         trade_offer_info = TradeBidOfferInfo(
             original_bid_rate=market_bid.original_price / market_bid.energy,
-            propagated_bid_rate=market_bid.energy_rate,
+            propagated_bid_rate=trade_original_info.propagated_bid_rate or market_bid.energy_rate,
```
Comment (hannesdiedrich) [inline src/gsy_e/models/market/grid_fees/constant_grid_fees.py:55]: This seems something unrelated to this PR. Can you please explain?

**Review by hannesdiedrich — COMMENTED** [review]
Two questions, once answered, LGTM

File: src/gsy_e/models/market/grid_fees/constant_grid_fees.py (hunk 3 above)
Comment (Hassan754) [reply src/gsy_e/models/market/grid_fees/constant_grid_fees.py:55]: Indeed, this is a bug that Luis correctly pointed out to, basically when we match with DOF, `market_bid.energy_rate` is not used, we use the rate from the requirement then (which is reflected in the  propagated_bid_rate field of the TradeBidOfferInfo)

File: src/gsy_e/models/market/two_sided.py (hunk 1 above)
Comment (Hassan754) [reply src/gsy_e/models/market/two_sided.py:451]: As far as I see, it won't be referenced anywhere so it should be deleted.

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Pull request merged**